ts/session/sending/MessageQueue.ts

```diff
@@ -1,54 +1,167 @@
-    throw new Error('Method not implemented.');
+
+  public async sendToGroup(message: OpenGroupMessage | ContentMessage): Promise<boolean> {
+    if (
+      !(message instanceof OpenGroupMessage) &&
+      !(message instanceof ClosedGroupMessage)
+    ) {
+      return false;
+    }
+
+    // Closed groups
+    if (message instanceof ClosedGroupMessage) {
+      // Get devices in closed group
+      const conversation = ConversationController.get(message.groupId);
+      const recipientsModels = conversation.contactCollection.models;
+      const recipients: Array<PubKey> = recipientsModels.map(
+        (recipient: any) => new PubKey(recipient.id)
+      );
+
+      await this.sendMessageToDevices(recipients, message);
+
+      return true;
+    }
+
+    // Open groups
+    if (message instanceof OpenGroupMessage) {
+      // No queue needed for Open Groups; send directly
+      await MessageSender.sendToOpenGroup(message);
+
+      return true;
+    }
+
+    return false;
   }
 
-  public processPending(device: string) {
-    // TODO: implement
+  public async sendSyncMessage(
+    message: ContentMessage,
+    sendTo: Array<PubKey>
+  ) {
+    // Sync with our devices
+    const promises = sendTo.map(async device => {
+      const syncMessage = await SyncMessageUtils.from(message, device);
+
+      return this.queue(device, syncMessage);
+    });
+
+    return Promise.all(promises);
   }
 
-  private processAllPending() {
-    // TODO: Get all devices which are pending here
+  public async processPending(device: PubKey) {
+    const messages = this.pendingMessageCache.getForDevice(device);
+
+    const hasSession = SessionProtocol.hasSession(device);
+    const conversation = ConversationController.get(device.key);
+    const isMediumGroup = conversation.isMediumGroup();
+
+    if (!isMediumGroup && !hasSession) {
+      await SessionProtocol.sendSessionRequestIfNeeded(device);
+
+      return;
+    }
+
+    const jobQueue = this.getJobQueue(device);
+    messages.forEach(message => {
+      if (!jobQueue.has(message.identifier)) {
+        const promise = jobQueue.add(async () => MessageSender.send(message));
+
+        promise
+          .then(() => {
+            // Message sent; remove from cache
+            void this.pendingMessageCache.remove(message);
+          })
+          // Message failed to send
+          .catch(() => null);
+      }
+    });
   }
 
-  private queue(device: string, message: ContentMessage) {
-    // TODO: implement
+  private async processAllPending() {
+    const devices = this.pendingMessageCache.getDevices();
+    const promises = devices.map(async device => this.processPending(device));
+
+    return Promise.all(promises);
   }
 
-  private queueOpenGroupMessage(message: OpenGroupMessage) {
-    // TODO: Do we need to queue open group messages?
-    // If so we can get open group job queue and add the send job here
+  private async queue(device: PubKey, message: ContentMessage) {
```

Comment:
```suggestion
  private async process(device: PubKey, message: ContentMessage) {
```
Change this so we don't get confused.